activemodel/lib/active_model/validator.rb

```diff
@@ -167,7 +168,9 @@ def check_validity!
 
     private
       def read_attribute_for_validation(record, attr_name)
-        record.read_attribute_for_validation(attr_name)
+        value = record.read_attribute_for_validation(attr_name)
+        throw(:allowed) if (value.nil? && options[:allow_nil]) || (value.blank? && options[:allow_blank])
```

Comment:
Moving the checks here actually make the API for custom validations worse. Users need to know that they need to deal with those options inside `read_attribute_for_validation` while before they didn't. I believe we should continue doing this check inside to keep users not needing to worry about checking for those options.

Can we split the job of `read_attribute_for_validation` in two? One which does `record.read_attribute_for_validation(attr_name)` and we use for the options checks and the other than calculate `value` that we will pass to `validate_each`? This will also avoid the `throw/abort`.

Reply:
I agree that it is not ideal, the intent of the implementation was to avoid extra `value` argument for `read_attribute_for_validation`.
I've given up that finally, splited the job and add `value` argument for `read_attribute_for_validation` in 720a60e68cf1945b6f486faca322d3acfa59c652 (perhaps `read_attribute_for_validation` no longer appropriate name now?).

Reply:
Agree the name is not matching the behaviour anymore. I'd call it `prepare_value_for_validation`.